### PR TITLE
CSHARP-3297: Handle incomplete regex patterns in JsonScanner

### PR DESCRIPTION
**Pull request by rkyser**

### CODE REVIEW

**Comment by rkyser** [comment]
Link to MongoDB Jira issue: https://jira.mongodb.org/projects/CSHARP/issues/CSHARP-3297

File: tests/MongoDB.Bson.Tests/IO/JsonScannerTests.cs [hunk 1]
```diff
@@ -405,6 +407,32 @@ public void TestMinusZeroExponentMinusTwelve()
...
+
+            // The completedTask should be the getNextTokenTask
+            Assert.Equal(getNextTokenTask, completedTask);
+
+            // Unwrap the FormatException from completedTask
+            var formatException = await Assert.ThrowsAsync<FormatException>(() => completedTask).ConfigureAwait(false);
+
+            Assert.StartsWith("Invalid JSON regular expression", formatException.Message);
+        }
+
```
Comment (rstam) [inline tests/MongoDB.Bson.Tests/IO/JsonScannerTests.cs:435]: I would name the test method something more specific than "invalid regular expression".

I think the test should be simplified to not have to deal with detecting infinite loops (now that the bug is fixed there is no infinite loop).

The test can be simply:

    [Theory]
    [InlineData("/")]
    [InlineData("/pattern")]
    [InlineData("/pattern\\/")]
    public void TestMissingClosingSlash(string jsonRegex)
    {
        var buffer = new JsonBuffer(jsonRegex);

        var exception = Record.Exception(() => JsonScanner.GetNextToken(buffer));

        exception.Should().BeOfType<FormatException>();
        exception.Message.Should().StartWith("Invalid JSON regular expression");
    }

Comment (rstam) [reply tests/MongoDB.Bson.Tests/IO/JsonScannerTests.cs:435]: Note: if we don't want to use FluentAssertions in a file that was first written before we started using FluentAssertions the last two lines can be rewritten using xUnit's `Assert`.

**Review by rstam — CHANGES_REQUESTED** [review]
Good catch!

File: tests/MongoDB.Bson.Tests/IO/JsonScannerTests.cs (hunk 1 above)
Comment (rkyser) [reply tests/MongoDB.Bson.Tests/IO/JsonScannerTests.cs:435]: Thank you for the review @rstam. I'll get these changes pushed in the next couple days.

Comment (rkyser) [reply tests/MongoDB.Bson.Tests/IO/JsonScannerTests.cs:423]: Changes have been pushed.

I opted to use the xUnit assertions to be consistent with the other tests in `JsonScannerTests`.

**Comment by rkyser** [comment]
@rstam Is there anything else you need from me? If so, let me know.

**Review by rstam — APPROVED** [review]
LGTM

Needs a second reviewer

File: tests/MongoDB.Bson.Tests/IO/JsonScannerTests.cs [hunk 2]
```diff
@@ -13,6 +13,8 @@
 * limitations under the License.
 */
 
+using System;
+using System.Threading.Tasks;
```
Comment (BorisDog) [inline tests/MongoDB.Bson.Tests/IO/JsonScannerTests.cs:17]: minor: looks like there is no need for `System.Threading.Tasks`

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Bson.Tests/IO/JsonScannerTests.cs (hunk 2 above)
Comment (rkyser) [reply tests/MongoDB.Bson.Tests/IO/JsonScannerTests.cs]: Good catch--change pushed.

**Comment by BorisDog** [comment]
LGTM
thanks!

**Review by BorisDog — APPROVED** [review]

**Comment by rstam** [comment]
Squashed, rebased and merged to master.

**Pull request closed without merging**